### PR TITLE
Updates GitHub Actions

### PR DESCRIPTION
**Pull request by appiepollo14**
https://github.com/MicroShed/microshed-testing/issues/297 Does this provide a solution for the first action checkbox? Tests for jaxrs and general are failing but could be fixed IMO.

### CODE REVIEW

**Comment by appiepollo14** [comment]
@KyleAure as you mention you can't build the project locally. Do you know whether the errors that occur in the GitHub CI after this PR are because of the CI or because of the code ?

File: .github/workflows/microshed-ci.yml [hunk 1]
```diff
@@ -17,14 +18,22 @@ env:
...
+      matrix:
+        os: [ ubuntu-latest ]
+        java: [ 11 ]
+      fail-fast: false
+      max-parallel: 4
+    name: Build ${{ matrix.java }}, ${{ matrix.os }}
+
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:31]: We should replace this with a concrete version number for security
```suggestion
    - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
```

File: .github/workflows/microshed-ci.yml [hunk 2]
```diff
@@ -17,14 +18,22 @@ env:
...
     - name: Set up Java
-      uses: joschi/setup-jdk@v2
+      uses: actions/setup-java@v3
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:33]: We should replace this with a concrete version number for security
```suggestion
      uses: actions/setup-java@0ab4596768b603586c0de567f2430c30f5b0d2b0 # v3.13.0
```

File: .github/workflows/microshed-ci.yml [hunk 3]
```diff
@@ -17,14 +18,22 @@ env:
...
       with:
         java-version: 11
-        openjdk_impl: openj9
+        distribution: 'zulu'
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:36]: The setup-java action does caching automatically now so we should switch this to: 
```suggestion
        distribution: 'temurin'
        cache: gradle
```

I'd like to stick with using either `temurin` or `semeru` as a distribution for this build.

File: .github/workflows/microshed-ci.yml [hunk 4]
```diff
@@ -17,14 +18,22 @@ env:
...
+    name: Build ${{ matrix.java }}, ${{ matrix.os }}
+
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
     - name: Set up Java
-      uses: joschi/setup-jdk@v2
+      uses: actions/setup-java@v3
       with:
         java-version: 11
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:35]: Switch this to use the test matrix:
```suggestion
        java-version: ${{ matrix.java }}
```

File: .github/workflows/microshed-ci.yml [hunk 5]
```diff
@@ -17,14 +18,22 @@ env:
...
-        openjdk_impl: openj9
+        distribution: 'zulu'
     - name: Apply repository caches
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:37]: This step can be removed

File: .github/workflows/microshed-ci.yml [hunk 6]
```diff
@@ -71,54 +82,56 @@ jobs:
...
+        with:
+          path: |
+            ~/.gradle/caches/modules-2
+            !~/.gradle/caches/modules-2/org.microshed*
+            ~/.gradle/wrapper
+            ~/.m2/repository/
+            !~/.m2/repository/org/microshed/
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle') }}-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:102]: Same changes as above

File: .github/workflows/microshed-ci.yml [hunk 7]
```diff
@@ -71,54 +82,56 @@ jobs:
...
+        run: docker version
+      - name: Set up testcontainers for podman
+        run: |
+          echo ryuk.container.privileged=true > ~/.testcontainers.properties
+      - name: Login to Docker Hub
+        uses: redhat-actions/podman-login@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME  }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+          registry: docker.io
```
Comment (KyleAure) [inline .github/workflows/microshed-ci.yml:115]: I don't think these setup steps are necessary.  A docker daemon should already be running on the GitHub Actions machine.

**Review by KyleAure — CHANGES_REQUESTED** [review]
Thanks for putting this together.  I think overall this is a good starting point.

File: .github/workflows/microshed-ci.yml (hunk 1 above)
Comment (KyleAure) [reply .github/workflows/microshed-ci.yml:31]: Dependabot will take care of this for us in the future :)

**Comment by KyleAure** [comment]
> @KyleAure as you mention you can't build the project locally. Do you know whether the errors that occur in the GitHub CI after this PR are because of the CI or because of the code ?

I think when I worked on this earlier this year I did have some tests failing but not all of them. I think right now the GH Actions build is failing due to some of the podman setup you have which I do not believe is necessary

**Comment by appiepollo14** [comment]
@aguibert Feel free to alter off course. When I run the CI with the credentials in my account. Both general and jaxrsjson are failing. Is that the state which you remember?

The error is no exposed ports and than a nullpointer. Sounds familiar?

File: .github/workflows/microshed-ci.yml (hunk 7 above)
Comment (appiepollo14) [reply .github/workflows/microshed-ci.yml]: Without the podman setup: com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"No such image: testcontainers/ryuk:0.3.0"} is the error...?

**Comment by appiepollo14** [comment]
This remove testspace analytics upload. If wanted, please add back.

File: sample-apps/everything-app/Dockerfile [hunk 8]
```diff
@@ -1,17 +1,3 @@
...
-#ADD build/libs/myservice.war /opt/jboss/wildfly/standalone/deployments/
-
-# Payara
-#FROM payara/micro:5.193
-#CMD ["--deploymentDir", "/opt/payara/deployments", "--noCluster"]
-#ADD build/libs/myservice.war /opt/payara/deployments
-
-# TomEE (not working yet)
-#FROM tomee:8-jre-8.0.0-M2-microprofile
-#COPY build/libs/myservice.war /usr/local/tomee/webapps/
```
Comment (KyleAure) [inline sample-apps/everything-app/Dockerfile:17]: I would keep these comments.  They allow users of different application servers a starting point to test out MicroshedTesting with their applications.

**Review by KyleAure — CHANGES_REQUESTED** [review]
I think this is really close!  Thanks for all the work you have done on this. Just some merge conflicts to resolve and one change request on my part :)

File: sample-apps/everything-app/Dockerfile (hunk 8 above)
Comment (appiepollo14) [reply sample-apps/everything-app/Dockerfile:17]: Each sample app has their own setup right? Why leave them in for all?

Comment (KyleAure) [reply sample-apps/everything-app/Dockerfile]: As in having a separate Dockerfile for each application server?

**Comment by appiepollo14** [comment]
@KyleAure is this mergable?

**Review by KyleAure — APPROVED** [review]

**Pull request merged**